src/renderer/dx/CustomTextLayout.cpp

```diff
@@ -398,36 +398,62 @@ CustomTextLayout::CustomTextLayout(gsl::not_null<IDWriteFactory1*> const factory
         // Walk through advances and space out characters that are too small to consume their box.
         for (auto i = run.glyphStart; i < (run.glyphStart + run.glyphCount); i++)
         {
-            // Advance is how wide in pixels the glyph is
-            auto& advance = _glyphAdvances.at(i);
-
-            // Offsets is how far to move the origin (in pixels) from where it is
-            auto& offset = _glyphOffsets.at(i);
+            const auto iClusterBegin = i;
 
             // Get how many columns we expected the glyph to have and multiply into pixels.
             UINT16 columns = 0;
             {
-                // Because of typographic features such as ligatures, it is well possible for a glyph to represent
-                //  multiple code points. Previous calls to IDWriteTextAnalyzer::GetGlyphs stores the mapping
+                // Because of typographic features such as ligatures, it is well possible for a cluster of glyph to
+                //  represent multiple code points. Previous calls to IDWriteTextAnalyzer::GetGlyphs stores the mapping
                 //  information between code points and glyphs in _glyphClusters.
-                // To properly allocate the columns for such glyphs, we need to find all characters that this glyph
-                //  is representing and add column counts for all the characters together.
+                // To properly allocate the columns for such glyph clusters, we need to find all characters that this
+                //  cluster is representing and add column counts for all the characters together.
+                // Inside a glyph cluster, we should try our best to respect the advances and offsets to keep glyphs properly
+                //  positioned. This means the glyph offset adjustment should be applied to all glyphs in cluster, and the
+                //  column-based advance adjustment should be applied on the last glyph of the cluster.
 
                 // Find the range for current glyph run in _glyphClusters.
                 const auto runStartIterator = _glyphClusters.begin() + run.textStart;
                 const auto runEndIterator = _glyphClusters.begin() + run.textStart + run.textLength;
 
                 // Find the range of characters that the current glyph is representing.
                 const auto firstIterator = std::find(runStartIterator, runEndIterator, i - run.glyphStart);

```

Comment:
These `find`s may impact performance.
Could we iterate through characters instead?

Reply:
I think considering that we are only searching within a single run, I personally think the impact should be pretty close to ignorable.